### ISSUE TITLE
[solc] ICE because "cannot convert magic variable super"

### ISSUE THREAD

**Issue by bshastry** [ethereum/solidity#9596]
## Description

```
contract C {
  C c;
  function h() public {
    super != c;
  }
}
```

throws

https://github.com/ethereum/solidity/blob/05901f5bc9607a6c3f0e4dd83954ebff17834eab/libsolidity/codegen/CompilerUtils.cpp#L754

## Environment

- Compiler version: latest develop

## Steps to Reproduce

```
$ solc test.sol
```

**Comment by hrkrshnn** [comment]
Maybe this will be fixed if we change super's type to TypeType. Relevant issue: https://github.com/ethereum/solidity/issues/7744

**Comment by cameel** [comment]
Another way to reproduce it is to compare `super` with `this`:
``` solidity
contract C {
    function h() public {
        this == super;
    }
}
```

I noticed that the assertion is triggered only if we're comparing with the current contract. If we compare with something else, we get a compilation error so I think the failure is not directly caused by `super` not being `TypeType`. Or at least there's already some workaround to deal with that problem and it's buggy:

```solidity
contract D {}
contract C {
    D d;
    function h() public {
        super != d;
    }
}
```
```
Error: Operator != not compatible with types contract super C and contract D
 --> super.sol:5:9:
  |
5 |         super != d;
  | 
```

Looks like no one is working on this bug right now so I'm going to take it.

**Comment by cameel** [comment]
Fixed in #9748. The cause was that any contract was considered implicitly convertible to the type of its `super`. When both appeared in the comparison operator the compiler tried to do just that, saw that a conversion is possible and and then proceeded to the conversion code where we assume that `super` won't ever appear.

So yeah, changing `super` so that it's a `TypeType` would fix it too but that's a bigger change and it's breaking. My fix just deals with the ICE.
